src/pjpersist/datamanager.py

```diff
@@ -853,7 +853,9 @@ def _begin(self, transaction):
         except psycopg2.Error as e:
             check_for_disconnect(e, 'PJDataManager._begin')
             raise
-        self._tpc_activated = True
+        finally:
+            if self._conn.status == psycopg2.extensions.STATUS_BEGIN:
+                self._tpc_activated = True
```

Comment:
I don't understand what this have to do with SystemExit.

Reply:
If SystemExit is raised in just the right time, `_tpc_activated` flag was not being set (because of exception) even though `tpc_begin` had just completed.  So when PJDataManager tried to clean up, it called `_conn.rollback()` - this is invalid call for a connection in status `STATUS_BEGIN`.  And so it failed with `ProgrammingError`

Reply:
SystemExit is just an exception we encountered in production, when a celery worker is being shut down while still trying to process some database task.

Reply:
You mean you think this happens when `SystemExit` is raised right between `tpc_begin()` and `_tpc_activated = True`?

Reply:
Yes.

Reply:
I see, this makes sense. but the code needs in addition:

1. Comment, explaining why this code is there
2. Logging message on warning level when we don't set _tpc_activated to True.

Reply:
@kedder re-arranged the code a bit and added logging.  Hope it reads better now.